File: csharp/yacht/Yacht.cs
using System;
using System.Linq;

public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{
    public static int Score(int[] dice, YachtCategory category)
    {
        switch (category)
        {
            case YachtCategory.Ones:
            case YachtCategory.Twos:
            case YachtCategory.Threes:
            case YachtCategory.Fours:
            case YachtCategory.Fives:
            case YachtCategory.Sixes:
                return dice.Where(x => x == (int)category).Sum();
            case YachtCategory.FullHouse:
                return IsFullHouse(dice) ? dice.Sum() : 0;
            case YachtCategory.FourOfAKind:
                return IsFourOfAKind(dice) ? GetFourOfAKindTotal(dice) : 0;
            case YachtCategory.LittleStraight:
                return IsLittleStraight(dice) ? 30 : 0;
            case YachtCategory.BigStraight:
                return IsBigStraight(dice) ? 30 : 0;
            case YachtCategory.Choice:
                return dice.Sum();
            case YachtCategory.Yacht:
                return dice.Distinct().Count() == 1 ? 50 : 0;
            default:
                return 0;
        }
    }

    private static bool IsFullHouse(int[] dice) => 
        dice.Distinct().Count() == 2 && (dice.Where(x => x == dice.Max()).Count() == 3 || dice.Where(x => x == dice.Min()).Count() == 3);

    private static bool IsFourOfAKind(int[] dice) => 
        dice.Where(x => x == dice.Max()).Count() >= 4 || dice.Where(x => x == dice.Min()).Count() >= 4;

    private static int GetFourOfAKindTotal(int[] dice) => 
        dice.FirstOrDefault(x => dice.Count(y => x == y) >= 4) * 4;

    private static bool IsLittleStraight(int[] dice) => 
        IsStraight(dice, 15);

    private static bool IsBigStraight(int[] dice) => 
        IsStraight(dice, 20);

    private static bool IsStraight(int[] dice, int sum) => 
        dice.Distinct().Count() == 5 && dice.Sum() == sum;
}

File: javascript/yacht/yacht.js
export const score = (rolls, category) => {
  switch (category) {
    case 'choice':
      return sum(rolls);
    case 'yacht':
      return rolls.every(r => r === rolls[0]) ? 50 : 0;
    case 'ones':
      return scoreNumber(rolls, 1);
    case 'twos':
      return scoreNumber(rolls, 2);
    case 'threes':
      return scoreNumber(rolls, 3);
    case 'fours':
      return scoreNumber(rolls, 4);
    case 'fives':
      return scoreNumber(rolls, 5);
    case 'sixes':
      return scoreNumber(rolls, 6);
    case 'full house':
      return scoreFullHouse(rolls);
    case 'four of a kind':
      return scoreFourOfAKind(rolls);
    case 'little straight':
    case 'big straight':
      return scoreStraight(rolls, category);
    default:
      break;
  }
};

const groupRolls = (rolls) => {
  return rolls.reduce((a, b) => (a[b] = a[b] + 1 || 1, a), {});
}

const scoreFourOfAKind = (rolls) => {
  let groups = groupRolls(rolls),
    index = Object.values(groups).findIndex(v => v >= 4);
  return index > -1 ? Object.keys(groups)[index] * 4 : 0;
}

const scoreFullHouse = (rolls) => {
  let groups = groupRolls(rolls),
    values = Object.values(groups);
  return values.findIndex(v => v === 3) > -1 && values.findIndex(v => v === 2) > -1 ? sum(rolls) : 0;
}

const scoreNumber = (rolls, number) => { 
  return rolls.filter(r => r === number).length * number;
}

const scoreStraight = (rolls, category)  => {
  if (rolls.length === new Set(rolls).size) {
    let s = sum(rolls);
    if ((s === 15 && category === 'little straight') || (s === 20 && category === 'big straight'))
      return 30;
  }
  return 0;
}

const sum = (rolls) => {
  return rolls.reduce((a, b) => a + b);
}